### PR TITLE
storage: support reverse scan

### PR DESCRIPTION
**Pull request by huachaohuang**
TiDB DDL needs to use this feature.

https://github.com/pingcap/kvproto/pull/258

### CODE REVIEW

File: src/storage/mod.rs [hunk 1]
```diff
@@ -657,23 +664,31 @@ impl Storage {
...
-                        .map_err(Error::from)
-                        .map(|results| {
-                            thread_ctx.collect_key_reads(CMD, results.len() as u64);
-                            results
-                                .into_iter()
-                                .map(|x| x.map_err(Error::from))
-                                .collect()
-                        })
+
+                    let scan_mode = if !options.reverse_scan {
```
Comment (breeswish) [inline src/storage/mod.rs]: I prefer avoiding double negative.. `if options.reverse_scan { ScanMode::Backword } ..`

File: src/storage/mod.rs [hunk 2]
```diff
@@ -657,23 +664,31 @@ impl Storage {
...
+                        ScanMode::Forward
+                    } else {
+                        ScanMode::Backward
+                    };
+
+                    let mut scanner = snap_store.scanner(scan_mode, options.key_only, None, None)?;
+                    let res = if !options.reverse_scan {
```
Comment (breeswish) [inline src/storage/mod.rs]: ditto

File: src/storage/mod.rs [hunk 3]
```diff
@@ -657,23 +664,31 @@ impl Storage {
...
+                        scanner.scan(start_key, limit)
+                    } else {
+                        scanner.reverse_scan(start_key, limit)
+                    };
+
+                    let statistics = scanner.get_statistics();
```
Comment (breeswish) [inline src/storage/mod.rs]: previous code only collect statistics when succeed

File: src/storage/mod.rs [hunk 4]
```diff
@@ -389,14 +389,21 @@ pub struct Options {
     pub lock_ttl: u64,
     pub skip_constraint_check: bool,
     pub key_only: bool,
+    pub reverse_scan: bool,
 }
 
 impl Options {
-    pub fn new(lock_ttl: u64, skip_constraint_check: bool, key_only: bool) -> Options {
+    pub fn new(
```
Comment (siddontang) [inline src/storage/mod.rs]: em, maybe we can provide a builder function to build Options now. Like

```
Options
pub fn new() -> Options
pub fn reverse_scan(self) -> Options
```

File: src/storage/mod.rs (hunk 3 above)
Comment (huachaohuang) [reply src/storage/mod.rs]: No, previous code collect statistics after scanning, no matter if `res` is ok or not.

**Comment by zhangjinpeng1987** [comment]
LGTM

**Comment by huachaohuang** [comment]
PTAL @breeswish

File: src/storage/mod.rs [hunk 5]
```diff
@@ -657,23 +664,31 @@ impl Storage {
...
+                    };
+
+                    let mut scanner = snap_store.scanner(scan_mode, options.key_only, None, None)?;
+                    let res = if options.reverse_scan {
+                        scanner.reverse_scan(start_key, limit)
+                    } else {
+                        scanner.scan(start_key, limit)
+                    };
+
+                    let statistics = scanner.get_statistics();
```
Comment (breeswish) [inline src/storage/mod.rs]: I'm still wondering this case. Previously, we have `collect_scan_count` and `collect_read_flow` in `snap_store.scanner(..).and_then(..)`, but now there is no `and_then`. There should be some differences in edge cases. The rest LGTM.

Comment (siddontang) [reply src/storage/mod.rs]: what is the difference?

**Comment by breeswish** [comment]
LGTM

File: src/storage/mod.rs (hunk 5 above)
Comment (breeswish) [reply src/storage/mod.rs]: When `snap_store.scanner(..)` returns `Err`, there will be differences. `and_then` only executes when it is `Ok()`.

Comment (huachaohuang) [reply src/storage/mod.rs]: The previous code did nothing when `snap_store.scanner()` returns `Err`, and now it returns directly when `snap_store.scanner()` returns `Err`, they are exactly the same.

**Pull request merged**